### PR TITLE
feature(Report to Compare Invoiced Items to received Stock)

### LINKED ISSUE #4380: Report to compare invoiced items to received stock
The patient financial activity report goes a long way to showing what the patient has been billed, what they have paid, and what they have received as stock.  However, for increased transparency, it would be nice to directly compare the billing and the invoicing.

Imagine the following scenario - the institution changes its policy for malaria cures.  Now, instead of 10 pills of quinine 300mg, it is 12 pills of quinine 300mg.  The pharmacist goes ahead and starts giving 12 pills to every patient, but the invoicing department didn't realize this and continues to bill for 10 pills total.  How does an administrator find out that what is being billed and what is being received are not the same thing?

I propose a new kind of report that attempts to join the inventory items billed on the stock consumed.  This would allow you to directly compare the items invoiced with the items distributed to the patient.  As a first design, it would look something like this:

image
_Fig 1:  Invoiced Items Compared to Stock Received_

Basically, for each invoice, we want to look up the stock items distributed to that invoice.  If the stock movement isn't linked to an invoice, we would show it in an "unassigned" table.

The table shows:
  1. The quantity of the item invoiced (from the invoice)
  2. The price of the item invoiced (from the invoice)
  3. The quantity of stock consumed (from the stock movement)
  4. The cost of the stock consumed (from the stock movement)
  5. The difference in quantity between the invoice and the stock movement.  If more stock is consumed than invoiced, we should highlight this in red.
  6. The difference in cost.  The we have consumed more value than we invoiced, we should highlight this in red.

### PR DESCRIPTION
**Pull request by lomamech**
- Implement the report to display what the patient has billed and what
  have received and what distribution without invoice Attibution

closes #4380

### CODE REVIEW

**Comment by kwilu** [comment]
Hello @lomamech, it seems like you have referenced #4380 in your pull request description, but you have not referenced them in your commit message description(s). When you reference an issue in a commit message, it automatically closes the corresponding issue when the commit is merged.

Please run `git commit --amend` in your command line client to amend your commit message description with `Fixes #4380.`.

An example of a correctly-formatted commit:
```
commit fabd5e450374c8dde65ec35f02140383940fe146
Author: kwilu
Date:   Sat Mar 18 13:42:40 2017 -0700

    pull requests: Check PR commits reference when issue is referenced.

    Fixes #51.
```

Thank you for your contributions to BHIMA!

**Comment by lomamech** [comment]
@jniles it's done
18UHA09KbM

File: server/models/migrations/next/migrate.sql [hunk 1]
```diff
@@ -1 +1,30 @@
...
+
+UPDATE stock_movement AS sm
+INNER JOIN (
+	SELECT sm.uuid, sm.document_uuid, sm.quantity, MONTH(sm.date) AS m_month, YEAR(sm.date) AS m_movement, p.id AS period_id
+	FROM stock_movement AS sm, period AS p
+	WHERE ((MONTH(p.start_date) = MONTH(sm.date)) AND (YEAR(p.start_date) = YEAR(sm.date)))
+) AS m_p ON sm.uuid = m_p.uuid
+SET sm.period_id = m_p.period_id;
+
+-- @lomamech 2020-04-29
```
Comment (jniles) [inline server/models/migrations/next/migrate.sql]: Everything in this file before this is in the previous migration script.  Can you delete everything above this and leave the new changes you made?

File: server/controllers/finance/reports/invoicedReceivedStock/report.handlebars [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+              <th style="width: 18%;" class="text-uppercase text-center">
+                {{translate 'TABLE.COLUMNS.VALUE'}}
+              </th>
+              <th style="width: 18%;" class="text-uppercase text-center">
+                {{translate 'TABLE.COLUMNS.TOTAL'}}
+              </th>
+            </tr>
+          </thead>
+          <tbody>
+            {{#each noInvoiceAttributionAggregat}}
```
Comment (jniles) [inline server/controllers/finance/reports/invoicedReceivedStock/report.handlebars]: If there isn't any extra stock not attributed to an invoice, can we hide this?

File: server/controllers/finance/reports/invoicedReceivedStock/index.js [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    iv.date, inv.text AS inventory_text, inv.uuid AS inventory_uuid, 
+    inv.code AS inventory_code, itm.quantity AS quantity_invoiced, itm.inventory_price AS price_invoiced
+    FROM patient AS p
+    JOIN debtor AS d ON d.uuid = p.debtor_uuid
+    JOIN invoice AS iv ON iv.debtor_uuid = d.uuid
+    JOIN invoice_item AS itm ON itm.invoice_uuid = iv.uuid
+    JOIN inventory AS inv ON inv.uuid = itm.inventory_uuid
+    JOIN document_map AS map ON map.uuid = iv.uuid
+    WHERE p.uuid = ? AND (DATE(iv.date) >= DATE(?) AND DATE(iv.date) <= DATE(?)) 
+    AND inv.consumable = 1
```
Comment (jniles) [inline server/controllers/finance/reports/invoicedReceivedStock/index.js:64]: :+1:

File: server/controllers/finance/reports/invoicedReceivedStock/index.js [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+const Patients = require('../../../medical/patients');
+
+const TEMPLATE = './server/controllers/finance/reports/invoicedReceivedStock/report.handlebars';
+// expose to the API
+exports.report = report;
+
+// default report parameters
+const DEFAULT_PARAMS = {
+  csvKey : 'invoiced_received_stock',
+  filename : 'REPORT.PATIENT_STANDING.DESCRIPTION',
```
Comment (jniles) [inline server/controllers/finance/reports/invoicedReceivedStock/index.js]: This filename should reflect what the report is.  The `filename` property is used to name the file when exporting to PDF.

File: server/controllers/finance/reports/invoicedReceivedStock/index.js [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+  orientation : 'landscape',
+};
+
+/**
+ * @function report
+ *
+ * @description
+ * This report makes it possible to attach the items billed to a patient on
+ * the articulated items consumed or distributed to the Patient.
```
Comment (jniles) [inline server/controllers/finance/reports/invoicedReceivedStock/index.js:22]: Good description :+1:

**Review by jniles — COMMENTED** [review]
@lomamech, you put this together really quickly!  Awesome.

This looks really good.  A few comments to make the module easier to use:
  1. Could you order the inventory items in each invoice alphabetically?  This would make it easier to compare when someone doesn't give Quinine 300mg, but gives Quinine 500mg.  
  2.  If there is no data for an inventory item or an invoice item, could you leave the cell blank instead of putting 0?  That way, the user can see easily what is missing.  I've noted that in red below.
  3. If an invoice has no consumable items and there is no stock linked to the invoice, we should hide it.  There is no point in showing it.
  4. If there are no extra stock items, we should hide the last table that shows extra stock items.  It's in red below.
  5.  Could you put a subtotal at the bottom of each $$ column in the report?

image

**Comment by lomamech** [comment]
seaiKbOf2c

**Comment by lomamech** [comment]
@jniles  it's done

**Comment by lomamech** [comment]
ujfOj35XCN

**Comment by lomamech** [comment]
@I just finished finalizing but also improving the report by correcting errors

File: server/models/migrations/next/migrate.sql (hunk 1 above)
Comment (jniles) [reply server/models/migrations/next/migrate.sql]: This still needs to be fixed.

**Comment by lomamech** [comment]
@jniles  All checks have passed

File: server/controllers/finance/reports/invoicedReceivedStock/index.js [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+const TEMPLATE = './server/controllers/finance/reports/invoicedReceivedStock/report.handlebars';
+// expose to the API
+exports.report = report;
+
+// default report parameters
+const DEFAULT_PARAMS = {
+  csvKey : 'invoiced_received_stock',
+  filename : 'REPORT.COMPARE_INVOICED_RECEIVED.TITLE',
+  orientation : 'landscape',
```
Comment (jniles) [inline server/controllers/finance/reports/invoicedReceivedStock/index.js:14]: I think this report makes more sense in a `portrait` than landscape.  Here is an example of my tests.

image

File: server/controllers/finance/reports/invoicedReceivedStock/index.js [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    SELECT map.text AS reference, inv.uuid AS inventory_uuid, inv.text AS inventory_text,
+    0 AS quantity, 0 AS price_invoiced, sm.invoice_uuid, sm.date,
+    SUM(sm.quantity) AS quantity_distributed, sm.unit_cost AS cost_distributed
+    FROM stock_movement AS sm
+    JOIN patient AS p ON p.uuid = sm.entity_uuid
+    JOIN lot AS l ON l.uuid = sm.lot_uuid
+    JOIN inventory AS inv ON inv.uuid = l.inventory_uuid
+    JOIN invoice AS iv ON iv.uuid = sm.invoice_uuid
+    JOIN document_map AS map ON map.uuid = iv.uuid
+    WHERE sm.invoice_uuid IS NOT NULL AND
```
Comment (jniles) [inline server/controllers/finance/reports/invoicedReceivedStock/index.js:75]: Do you need this `sm.invoice_uuid IS NOT NULL`?  I think the `JOIN` on inventory should already take care of that, no?

File: server/controllers/finance/reports/invoicedReceivedStock/report.handlebars [hunk 8]
```diff
@@ -0,0 +1,176 @@
+<!doctype html>
+<html>
+  {{> head title="REPORT.COMPARE_INVOICED_RECEIVED.TITLE"}}
+  <body>
+  <div class="container">
```
Comment (jniles) [inline server/controllers/finance/reports/invoicedReceivedStock/report.handlebars]: Can you make this `container-fluid` to not have as much padding?

File: server/controllers/finance/reports/invoicedReceivedStock/report.handlebars [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+              <td class="text-right"> {{ debcred unit_cost ../../metadata.enterprise.currency_id }} </td>
+              <td class="text-right"> <strong>{{ debcred total_cost ../../metadata.enterprise.currency_id }}</strong> </td>
+            </tr>
+          {{/each}}
+          <tr style="background-color:#efefef;">
+            <td colspan="4" class="text-right"><strong>{{translate 'TABLE.COLUMNS.TOTAL'}}</strong></td>
+            <td class="text-right"><strong> {{ debcred total_movement ../../metadata.enterprise.currency_id }}  </strong></td>
+          </tr>
+          <tr>
+            <td colspan="5"></td>
```
Comment (jniles) [inline server/controllers/finance/reports/invoicedReceivedStock/report.handlebars]: Why do you have these extra `<td>`s?  They add a kind of weird three lines below the report.  Could you just put a `padding` or `margin` on something to add space if you need it?  Or even a `<br>`?

image

File: server/models/migrations/next/migrate.sql [hunk 10]
```diff
@@ -8,4 +8,34 @@ INSERT INTO unit VALUES
...
   (15, 'STOCK_FLUX.INVENTORY_ADJUSTMENT');
-  
\ No newline at end of file
+  
+/* migrate next */
+
+-- @jeremielodi 23-04-2020
+-- stock consumption graph report
+
+INSERT INTO unit VALUES
```
Comment (jniles) [inline server/models/migrations/next/migrate.sql]: The only thing that should be in this file is the `INSERT INTO flux` from @mbayopanda and your additions:
```sql
INSERT INTO unit VALUES
  (270, 'compare invoiced to received','TREE.COMPARE_INVOICED_RECEIVED','Compare invoiced items to received stock', 144,'/modules/reports/invoicedReceivedStock','/reports/invoicedReceivedStock');

INSERT INTO `report` (`report_key`, `title_key`) VALUES
  ('invoicedReceivedStock', 'REPORT.COMPARE_INVOICED_RECEIVED.TITLE');
```

Please make these changes.

**Review by jniles — COMMENTED** [review]
@lomamech this is really close!  I've tested it on live data and it works well, the best I can tell.

I've made a few comments.  The `migration/next.sql` is really important to address before a merge.  It will break upgrades if we don't address it.

File: server/controllers/finance/reports/invoicedReceivedStock/index.js (hunk 6 above)
Comment (jniles) [reply server/controllers/finance/reports/invoicedReceivedStock/index.js:14]: Or maybe just removing the padding would address this, actually.

Comment (lomamech) [reply server/controllers/finance/reports/invoicedReceivedStock/index.js:14]: > Or maybe just removing the padding would address this, actually.

I also just noticed that when you download a report, the shades of gray disappear, with you noticing this problem 
image

Comment (jniles) [reply server/controllers/finance/reports/invoicedReceivedStock/index.js:14]: Yeah, that has to do with #4382.  It's why I was hoping for a solution that didn't involve `!important;` everywhere.  Don't worry about that for now.

**Comment by lomamech** [comment]
@jniles it's done

**Review by jniles — APPROVED** [review]
This works really well.  Thanks @lomamech !

bors r+

**Pull request merged**